alicloud/resource_alicloud_cs_kubernetes.go

```diff
@@ -603,32 +744,74 @@ func buildKunernetesArgs(d *schema.ResourceData, meta interface{}) (*cs.Kubernet
 		ServiceCIDR:              d.Get("service_cidr").(string),
 		SSHFlags:                 d.Get("enable_ssh").(bool),
-		ImageID:                  KubernetesImageId,
 		CloudMonitorFlags:        d.Get("install_cloud_monitor").(bool),
 		ZoneId:                   zoneId,
+	}, nil
+}
+
+func buildKubernetesMultiAZArgs(d *schema.ResourceData, meta interface{}) (*cs.KubernetesMultiAZCreationArgs, error) {
+	client := meta.(*AliyunClient)
+
+	// Ensure instance_type is valid
+	zoneId, validZones, err := client.DescribeAvailableResources(d, meta, InstanceTypeResource)
+	if err != nil {
+		return nil, err
 	}
+	instanceTypes := expandStringList(d.Get("master_instance_types").([]interface{}))
+	instanceTypes = append(instanceTypes, expandStringList(d.Get("worker_instance_types").([]interface{}))...)
 
-	if v, ok := d.GetOk("vswitch_id"); ok && len(Trim(v.(string))) > 0 {
-		stackArgs.VSwitchID = Trim(v.(string))
-		vsw, err := client.DescribeVswitch(stackArgs.VSwitchID)
-		if err != nil {
+	for _, instanceType := range instanceTypes {
+		if err := meta.(*AliyunClient).InstanceTypeValidation(instanceType, zoneId, validZones); err != nil {
 			return nil, err
 		}
-		stackArgs.VPCID = vsw.VpcId
-		if stackArgs.ZoneId != "" && vsw.ZoneId != vsw.ZoneId {
-			return nil, fmt.Errorf("The specified vswitch %s isn't in the zone %s.", vsw.VSwitchId, stackArgs.ZoneId)
-		}
-		stackArgs.ZoneId = vsw.ZoneId
-	} else if !stackArgs.SNatEntry {
-		return nil, fmt.Errorf("The automatic created VPC and VSwitch must set 'new_nat_gateway' to 'true'.")
 	}
 
-	return &cs.KubernetesCreationArgs{
-		Name:              clusterName,
-		ClusterType:       "Kubernetes",
-		DisableRollback:   true,
-		TimeoutMins:       60,
-		KubernetesVersion: stackArgs.KubernetesVersion,
-		StackParams:       *stackArgs,
+	var clusterName string
+	if v, ok := d.GetOk("name"); ok {
+		clusterName = v.(string)
+	} else {
+		return nil, errors.New("The 'name' is required for Kubernetes MultiAZ.")
+	}
+
+	masterInstanceTypes := expandStringList(d.Get("master_instance_types").([]interface{}))
+	workerInstanceTypes := expandStringList(d.Get("worker_instance_types").([]interface{}))
+	vswitchIDs := expandStringList(d.Get("vswitch_ids").([]interface{}))
+	workerNumbers := expandIntList(d.Get("worker_numbers").([]interface{}))
+
+	vsw, err := client.DescribeVswitch(vswitchIDs[0])
+	if err != nil {
+		return nil, err
+	}
+
+	return &cs.KubernetesMultiAZCreationArgs{
+		Name:                     clusterName,
+		ClusterType:              "Kubernetes",
+		DisableRollback:          true,
+		TimeoutMins:              60,
+		MultiAZ:                  true,
+		MasterInstanceTypeA:      masterInstanceTypes[0],
+		MasterInstanceTypeB:      masterInstanceTypes[1],
+		MasterInstanceTypeC:      masterInstanceTypes[2],
+		WorkerInstanceTypeA:      workerInstanceTypes[0],
+		WorkerInstanceTypeB:      workerInstanceTypes[1],
+		WorkerInstanceTypeC:      workerInstanceTypes[2],
```

Comment:
I think there should add 3 conditions to ensure the length of each masterInstanceTypes, workerInstanceTypes and vswitchIDs is 3.

Reply:
@xiaozhu36 those check has been implements in the isMultiAZClusterAndCheck method.